src/JuliusSweetland.OptiKey.Core/Services/KeyboardOutputService.cs

```diff
@@ -1055,7 +1071,14 @@ private void PublishKeyPress(char character)
                 {
                     Log.InfoFormat("Publishing '{0}' => as virtual key code {1} (using hard coded mapping)",
                         character.ToPrintableString(), virtualKeyCode);
-                    publishService.KeyDownUp(virtualKeyCode.Value);
+                    if (type == KeyPressKeyValue.KeyPressType.Press)
+                        publishService.KeyDown(virtualKeyCode.Value);
+                    else if (type == KeyPressKeyValue.KeyPressType.Release)
+                        publishService.KeyUp(virtualKeyCode.Value);
+                    else if (type == KeyPressKeyValue.KeyPressType.Toggle)
+                        publishService.KeyToggle(virtualKeyCode.Value);
+                    else
+                        publishService.KeyDownUp(virtualKeyCode.Value);
                     return;
                 }
 
```

Comment:
If Settings.Default.PublishVirtualKeyCodesForCharacters == false then toggling/pressing/releasing keys will not work correctly (except for these keys: https://sourcegraph.com/github.com/OptiKey/OptiKey@f5b394472cdff2fe1e020dde1b6ee011e15fd8b9/-/blob/src/JuliusSweetland.OptiKey.Core/Extensions/FunctionKeysExtensions.cs#L16:39). You might want to think about what the correct behaviour is here - if we receive a request to KeyDown/KeyDown/Toggle a character or key and this setting is false then we will ignore that and send the text to the OS in a similar way as if a key down&up had been completed. Either we could add extra checks here and only allow the sending of text if this setting is false AND we are performing a PressAndRelease, otherwise ALWAYS attempt to lookup the virtual key code, or we live with this setting overriding all key behaviour to output text. I would lean towards only sending text if we are performing a PressAndRelease, but log the fact that we are ignoring this setting for Press, Release, and Toggle so at least the log is clear that the settings dictate we send text, but that it makes no sense in this situation.

Reply:
As noted in another comment. I think we'll make KeyDown/Up/Toggle special.